### ISSUE TITLE
When multipleHits is true, remove hints that have been clicked

### ISSUE THREAD

**Issue by dufferzafar** [brookhong/Surfingkeys#1240]
```
mapkey('cf', '#1Open multiple links in a new tab', function() {
    Hints.create("", Hints.dispatchMouseClick, {multipleHits: true});
});
```

When I press `cf`, I can quickly open a bunch of links, but it'd be nice if the hints would get removed from links that have already been clicked - this will reduce the "visual noise" on the screen.

/cc @b0o Wondering if you might be aware of how this could be done?

In the event handler, could we find out the hints dom node and remove it?

If that happens, then other hints won't need to be recomputed.

**Comment by b0o** [comment]
Good idea, I agree it would be a nice UX improvement.

Preliminary thoughts:

I don't think removing the DOM node for a clicked hint would be enough, since the nodes are destroyed and re-created on scroll. I think we would need to keep track of the actual element the hint was created for.

We could keep track of handled hints [here](https://github.com/brookhong/Surfingkeys/blob/4e6cab8ce2a3eff23841fabcfa5cad8e58950c06/content_scripts/hints.js#L116) before calling `refresh()`.

We could append the DOM node referenced by `matches[0]` to an array, e.g. `handledHints.append(matches[0])` - the array `handledHints` would be local to the outer `createHints()` function scope. Then inside `refresh()` we could filter out any handled nodes. Finally inside `hide()` we would probably want to reset the array.

I question whether the hint should be fully removed or maybe instead have its opacity lowered? I'm not sure if it would feel jarring/unexpected to the user. I feel like it would have to be tested to see what works better.

**Comment by Shamaoke** [comment]
I think, the current behavior shouldn't be changed. Here's why:

• In most cases there's no any “visual noise”. In cases there is, it's a “positive visual noise” so to say. That is, what you see is needed for you, so I hard to call it “visual noise” at all. Moreover, removing a couple of hints will not affect it in any way. Just check: go to [old.reddit.com](https://old.reddit.com), call hints, remove 5..10 of them, you'll see, that there will be no any substantial difference on “visual noise”. Also, there's the `Ctrl+Space` hotkey you can use to temporary clear the screen out of hints;

• Besides its function as controls, hints also serves as indicators of where clickable elements are on a page. You won't get a hint what are these elements if you will be removing hint markers;

• In some use cases, It often needs to open the same page multiple times. Removing hints will not allow one to do it;

• In searching forms you should be able to activate control elements multiple times (e. g. check and un-check checkboxes). Stripping hints will completely break the workflow with forms.

In general, I think there's no any problem with hints staying on a page. And to be honest, I don't understand the whole motivation behind this proposal. Moreover, I think that implementing this feature will likely create user experience issues in the future.

**Comment by dufferzafar** [comment]
@Shamaoke You've raised some valid points, which is why this is not going to change the default behaviour in any way!

I created this issue here because I'm new to surfing keys and wasn't even sure if there was already a way to do this. And as I see it, the entire point of Surfing Keys being so customizable is that people can have their own config. 

So for me, when I go on old.reddit.com and call hints, I only see very few hints (because I've customized that) and I _personally_ would like to remove the ones that I've already clicked on.

But as you said, this won't work in a lot of cases, which is why, the default behaviour should stay.

**Comment by brookhong** [comment]
I believe that here we are talking about only the short period of a triggering of `cf`, as excluding nodes or re-styling hints for clicked nodes does not make sense for another call of Hints.

In one call of Hints, re-calculating will be involved if the clicked hints were removed, and labels will be re-generated, a frequently changing labels are not user friendly. Maybe using a different style for clicked hints is better.

Answer to another question here is NO. We currently have no settings to exclude elements from Hints.

Thanks for the question, but we still need more votes on these 2 requirements.

* different styles for clicked hints in one call of `cf`
* new settings to exclude unwanted elements from Hints

**Comment by b0o** [comment]
> I believe that here we are talking about only the short period of a triggering of cf, as excluding nodes or re-styling hints for clicked nodes does not make sense for another call of Hints.

This is how I interpret the original suggestion. I am not sure if @Shamaoke is also on the same page, or is referring to hiding hints between separate invocations of `cf`, which I agree would be undesirable.

I like the idea of signifying hints that've been visited during the current invocation, but not necessarily hiding them. Almost every time I use the `multipleHits` feature to open lots of links, I lose track of which ones I've clicked.

The basic functionality of hiding visited hints is already possible to implement in one's config without any changes to SurfingKeys itself - a while back I implemented a [PR](https://github.com/brookhong/Surfingkeys/pull/852) that allows the creation of hints via an array of DOM nodes. Using this, we can manually run the query selector and filter out visited elems:

```js
const createHintsMultipleHits = (selector, visited = []) => {
  const elems = Array.from(document.querySelectorAll(selector))
    .filter((e) => !visited.includes(e))
  Hints.create(elems, (res) => {
    RUNTIME("openLink", { tab: { tabbed: true, active: false }, url: res.href })
    // setTimeout is needed to ensure that the next call to Hints.create() occurs after the
    // first one has returned
    setTimeout(() => createHintsMultipleHits(selector, visited.concat([res])), 1)
  })
}

mapkey("AA", "MultipleHits", () => createHintsMultipleHits("a"))
```

**Comment by Shamaoke** [comment]
> Screencast of this in action.

How to open the same link second time in the same invocation of `cf`?

Try to check all the check boxes in [this](https://jsfiddle.net/u69wdsgy) example, then try to un-check one (or all) of them. How to do it in the same invocation of `cf`?

**Comment by b0o** [comment]
@Shamaoke Yes I agree it’s not ideal in all cases. I still prefer the idea of lowering the opacity or changing the color of visited hints rather than hiding them, that was you can always click on a hint multiple times if you want. I just wanted to demonstrate how one can accomplish this now in their own config without any changes to SurfingKeys itself.
